File: comphelper/source/misc/officeresourcebundle.cxx
/* -*- Mode: C++; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- */
/*
 * This file is part of the LibreOffice project.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 *
 * This file incorporates work covered by the following license notice:
 *
 *   Licensed to the Apache Software Foundation (ASF) under one or more
 *   contributor license agreements. See the NOTICE file distributed
 *   with this work for additional information regarding copyright
 *   ownership. The ASF licenses this file to you under the Apache
 *   License, Version 2.0 (the "License"); you may not use this file
 *   except in compliance with the License. You may obtain a copy of
 *   the License at http://www.apache.org/licenses/LICENSE-2.0 .
 */

#include <comphelper/officeresourcebundle.hxx>

#include <com/sun/star/resource/MissingResourceException.hpp>
#include <com/sun/star/resource/XResourceBundle.hpp>
#include <com/sun/star/resource/XResourceBundleLoader.hpp>
#include <osl/mutex.hxx>
#include <osl/diagnose.h>
#include <rtl/ustrbuf.hxx>


namespace comphelper
{


    using ::com::sun::star::uno::Reference;
    using com::sun::star::resource::XResourceBundle;
    using com::sun::star::resource::XResourceBundleLoader;
    using com::sun::star::resource::MissingResourceException;
    using ::com::sun::star::uno::XComponentContext;
    using ::com::sun::star::uno::Exception;
    using ::com::sun::star::uno::Any;

    class ResourceBundle_Impl
    {
    private:
        Reference< XComponentContext >  m_xContext;
        OUString                 m_sBaseName;
        Reference< XResourceBundle >    m_xBundle;
        bool                            m_bAttemptedCreate;
        mutable ::osl::Mutex            m_aMutex;

    public:
        ResourceBundle_Impl( const Reference< XComponentContext >& _context, const OUString& _baseName )
            :m_xContext( _context )
            ,m_sBaseName( _baseName )
            ,m_bAttemptedCreate( false )
        {
        }

    public:
        /** loads the string with the given resource id from the resource bundle
            @param  _resourceId
                the id of the string to load
            @return
                the requested resource string. If no string with the given id exists in the resource bundle,
                an empty string is returned. In a non-product version, an OSL_ENSURE will notify you of this
                then.
        */
        OUString loadString( sal_Int32 _resourceId ) const;

        /** determines whether the resource bundle has a string with the given id
            @param  _resourceId
                the id of the string whose existence is to be checked
            @return
                <TRUE/> if and only if a string with the given ID exists in the resource
                bundle.
        */
        bool            hasString( sal_Int32 _resourceId ) const;

    private:
        /** loads the bundle represented by the instance

            The method is safe against multiple calls: If a previous call succeeded or failed, the
            previous result will be returned, without any other processing.

            @precond
                Our mutex is locked.
        */
        bool    impl_loadBundle_nothrow();

        /** returns the resource bundle key for a string with a given resource id
        */
        static OUString
                impl_getStringResourceKey( sal_Int32 _resourceId );
    };


    OUString ResourceBundle_Impl::impl_getStringResourceKey( sal_Int32 _resourceId )
    {
        OUStringBuffer key;
        key.append( "string:" );
        key.append( _resourceId );
        return key.makeStringAndClear();
    }


    OUString ResourceBundle_Impl::loadString( sal_Int32 _resourceId ) const
    {
        ::osl::MutexGuard aGuard( m_aMutex );

        OUString sString;

        if ( const_cast< ResourceBundle_Impl* >( this )->impl_loadBundle_nothrow() )
        {
            try
            {
                OSL_VERIFY( m_xBundle->getByName( impl_getStringResourceKey( _resourceId ) ) >>= sString );
            }
            catch( const Exception& )
            {
                OSL_FAIL( "ResourceBundle_Impl::loadString: caught an exception!" );
            }
        }
        return sString;
    }


    bool ResourceBundle_Impl::hasString( sal_Int32 _resourceId ) const
    {
        ::osl::MutexGuard aGuard( m_aMutex );

        bool has = false;

        if ( const_cast< ResourceBundle_Impl* >( this )->impl_loadBundle_nothrow() )
        {
            try
            {
                has = m_xBundle->hasByName( impl_getStringResourceKey( _resourceId ) );
            }
            catch( const Exception& )
            {
                OSL_FAIL( "ResourceBundle_Impl::hasString: caught an exception!" );
            }
        }
        return has;
    }


    bool ResourceBundle_Impl::impl_loadBundle_nothrow()
    {
        if ( m_bAttemptedCreate )
            return m_xBundle.is();

        m_bAttemptedCreate = true;

        Reference< XResourceBundleLoader > xLoader;
        try
        {
            Any aValue( m_xContext->getValueByName( "/singletons/com.sun.star.resource.OfficeResourceLoader" ) );
            OSL_VERIFY( aValue >>= xLoader );
        }
        catch( const Exception& )
        {
            OSL_FAIL( "ResourceBundle_Impl::impl_loadBundle_nopthrow: could not create the resource loader!" );
        }

        if ( !xLoader.is() )
            return false;

        try
        {
            m_xBundle = xLoader->loadBundle_Default( m_sBaseName );
        }
        catch( const MissingResourceException& )
        {
            OSL_FAIL( "ResourceBundle_Impl::impl_loadBundle_nopthrow: missing the given resource bundle!" );
        }

        return m_xBundle.is();
    }


    OfficeResourceBundle::OfficeResourceBundle( const Reference< XComponentContext >& _context, const sal_Char* _bundleBaseAsciiName )
        :m_pImpl( new ResourceBundle_Impl( _context, OUString::createFromAscii( _bundleBaseAsciiName ) ) )
    {
    }


    OfficeResourceBundle::~OfficeResourceBundle()
    {
    }


    OUString OfficeResourceBundle::loadString( sal_Int32 _resourceId ) const
    {
        return m_pImpl->loadString( _resourceId );
    }


    bool OfficeResourceBundle::hasString( sal_Int32 _resourceId ) const
    {
        return m_pImpl->hasString( _resourceId );
    }


} // namespace comphelper


/* vim:set shiftwidth=4 softtabstop=4 expandtab: */

File: connectivity/java/sdbc_postgresql/src/com/sun/star/sdbcx/comp/postgresql/sdbcx/OfficeResourceBundle.java
/**************************************************************
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 *
 *************************************************************/

package com.sun.star.sdbcx.comp.postgresql.sdbcx;

import com.sun.star.lang.NullPointerException;
import com.sun.star.resource.MissingResourceException;
import com.sun.star.resource.OfficeResourceLoader;
import com.sun.star.resource.XResourceBundle;
import com.sun.star.resource.XResourceBundleLoader;
import com.sun.star.sdbcx.comp.postgresql.comphelper.CompHelper;
import com.sun.star.uno.AnyConverter;
import com.sun.star.uno.XComponentContext;

public class OfficeResourceBundle implements AutoCloseable {

    private XComponentContext context;
    private String baseName;
    private boolean haveAttemptedCreate;
    private XResourceBundle bundle;


    /** constructs a resource bundle
        @param  context
            the component context to operate in
        @param  bundleBaseName
            the base name of the resource file which should be accessed (*without* the SUPD!)
        @raises ::com::sun::star::lang::NullPointerException
            if the given component context is <NULL/>
     */
    public OfficeResourceBundle(XComponentContext context, String bundleBaseName) throws NullPointerException {
        if (context == null) {
            throw new NullPointerException();
        }
        this.context = context;
        this.baseName = bundleBaseName;
        this.haveAttemptedCreate = false;
    }

    @Override
    public void close() {
        CompHelper.disposeComponent(bundle);
    }

    /** loads the string with the given resource id from the resource bundle
        @param  _resourceId
            the id of the string to load
        @return
            the requested resource string. If no string with the given id exists in the resource bundle,
            an empty string is returned. In a non-product version, an OSL_ENSURE will notify you of this
            then.
    */
    public String loadString( int _resourceId ) {
        synchronized (this) {
            String ret = "";
            if (loadBundle()) {
                try {
                    Object value = bundle.getByName(getStringResourceKey(_resourceId));
                    ret = AnyConverter.toString(value);
                } catch (com.sun.star.uno.Exception ex) {
                }
            }
            return ret;
        }
    }

    /** determines whether the resource bundle has a string with the given id
        @param  _resourceId
            the id of the string whose existence is to be checked
        @return
            <TRUE/> if and only if a string with the given ID exists in the resource
            bundle.
    */
    public boolean hasString( int _resourceId ) {
        synchronized (this) {
            boolean ret = false;
            if (loadBundle()) {
                ret = bundle.hasByName(getStringResourceKey(_resourceId));
            }
            return ret;
        }
    }

    private String getStringResourceKey(int resourceId) {
        return "string:" + resourceId;
    }

    private boolean loadBundle() {
        if (haveAttemptedCreate) {
            return bundle != null;
        }
        haveAttemptedCreate = true;

        try {
            XResourceBundleLoader loader = OfficeResourceLoader.get(context);
            try {
                if (loader == null) {
                    return false;
                }
                try {
                    bundle = loader.loadBundle_Default(baseName);
                } catch (MissingResourceException missingResourceException) {
                }
                return bundle != null;
            } finally {
                CompHelper.disposeComponent(loader);
            }
        } catch (Exception exception) {
        }
        return false;
    }
}
